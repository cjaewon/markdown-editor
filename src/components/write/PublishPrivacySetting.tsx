import * as React from 'react';
import styled, { css } from 'styled-components';
import PublishSection from './PublishSection';
import palette from '../../lib/styles/palette';
import { GlobeIcon, LockIcon } from '../../static/svg';

const PublishPrivacySettingBlock = styled(PublishSection)`
  outline: none;
  border: none;
  .contents {
    display: flex;
  }
`;

const Button = styled.button<{ active: boolean }>`
  outline: none;
  border: none;
  flex: 1;
  height: 3rem;
  display: inline-flex;
  align-items: center;
  justify-content: flex-start;
  font-weight: bold;
  background: white;
  font-size: 1.125rem;
  color: ${palette.gray6};
  box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.05);
  padding: 0;
  padding-left: 1rem;
  border-radius: 4px;
  cursor: pointer;
  border: solid 1px transparent;
  &:hover {
    background: #fdfdfd;
  }
  ${props =>
    props.active &&
    css`
      border: solid 1px ${palette.teal5};
      color: ${palette.teal5};
    `}

  svg {
  }
  & + & {
    margin-left: 1rem;
  }
  .description {
    flex: 1;
    display: flex;
    justify-content: center;
    align-items: center;
  }
`;

export interface PublishPrivacySettingProps {
  isPrivate: boolean;
  onSelect: (value: boolean) => any;
}

const { useCallback } = React;

const PublishPrivacySetting: React.FC<PublishPrivacySettingProps> = ({
  isPrivate,
  onSelect,
}) => {
  const onClickPrivate = useCallback(() => {
    onSelect(true);
  }, [onSelect]);
  const onClickPublic = useCallback(() => {
    onSelect(false);
  }, [onSelect]);
  return (
    <PublishPrivacySettingBlock title="공개 설정">
      <Button active={!isPrivate} onClick={onClickPublic}>
        <GlobeIcon />
        <div className="description">전체 공개</div>
      </Button>
      <Button active={isPrivate} onClick={onClickPrivate}>
        <LockIcon />
        <div className="description">비공개</div>
      </Button>
    </PublishPrivacySettingBlock>
  );
};

export default PublishPrivacySetting;
