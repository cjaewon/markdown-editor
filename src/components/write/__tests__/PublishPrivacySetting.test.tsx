import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PublishPrivacySetting, {
  PublishPrivacySettingProps,
} from '../PublishPrivacySetting';

describe('PublishPrivacySetting', () => {
  const setup = (props: Partial<PublishPrivacySettingProps> = {}) => {
    const initialProps: PublishPrivacySettingProps = {
      isPrivate: false,
      onSelect: () => {},
    };
    const utils = render(
      <PublishPrivacySetting {...initialProps} {...props} />,
    );
    const buttons = {
      public: utils.getByText('전체 공개').parentElement,
      private: utils.getByText('비공개').parentElement,
    };
    return {
      ...utils,
      buttons,
    };
  };
  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
  describe('shows active style', () => {
    it('PUBLIC', () => {
      const utils = setup();
      expect(getComputedStyle(utils.buttons.public).color).toBe(
        'rgb(32, 201, 151)',
      );
    });
    it('PRIVATE', () => {
      const utils = setup({ isPrivate: true });
      expect(getComputedStyle(utils.buttons.private).color).toBe(
        'rgb(32, 201, 151)',
      );
    });
  });
  it('calls onSelect function onClick', () => {
    const onSelect = jest.fn();
    const utils = setup({ onSelect });
    fireEvent.click(utils.buttons.private);
    expect(onSelect).toBeCalledWith(true);
    fireEvent.click(utils.buttons.public);
    expect(onSelect).toBeCalledWith(false);
  });
});
