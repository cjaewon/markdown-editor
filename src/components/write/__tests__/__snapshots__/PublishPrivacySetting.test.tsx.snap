// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublishPrivacySetting matches snapshot 1`] = `
<div>
  <section
    class="sc-bdVaJa ezdyV sc-bwzfXH difkFg"
  >
    <h3>
      공개 설정
    </h3>
    <div
      class="contents"
    >
      <button
        class="sc-htpNat iAsitD"
      >
        <svg>
          icon-globe.svg
        </svg>
        <div
          class="description"
        >
          전체 공개
        </div>
      </button>
      <button
        class="sc-htpNat gjoXMo"
      >
        <svg>
          icon-lock.svg
        </svg>
        <div
          class="description"
        >
          비공개
        </div>
      </button>
    </div>
  </section>
</div>
`;
